### PR TITLE
Eventually consistent update/delete

### PR DESCRIPTION
**Pull request by xbrianh**
Update and delete events are recorded as markers but not applied. Markers are applied during an offline process. This allows safe distributed updates and deletes.

### CODE REVIEW

File: flashflood/__init__.py [hunk 1]
```diff
@@ -229,14 +274,10 @@ def _journal_ids(self, from_date=None, to_date=None):
...
-    def _delete_all_journals(self):
-        journal_ids = [journal_id
-                       for journal_id in self._journal_ids()]
-        self._delete_journals(journal_ids)
-
     def _destroy(self):
-        for item in self.bucket.objects.filter(Prefix=self.root_prefix):
-            item.delete()
+        with S3Deleter(self.bucket) as s3d:
+            for item in self.bucket.objects.filter(Prefix=self.root_prefix):
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply flashflood/__init__.py]: It's a good idea, but that information is not available to the `flash-flood` library. Those safety nets need to be applied to the data-store: we should take care that flashflood is never pointed at the main buckets of the DSS.

Incidentally, the main buckets of the data-store deny deletes privileges to almost everything that's [not and admin or Kharon](https://github.com/HumanCellAtlas/dcp-infra/blob/master/terraform/accounts/humancellatlas/policy_aws_replica.tf).

File: flashflood/__init__.py [hunk 2]
```diff
@@ -52,22 +52,91 @@ def __init__(self, s3_resource, bucket, root_prefix):
...
+            journal_id = _JournalID.make(timestamp, None, blob_id)
+            manifest = dict(journal_id=journal_id,
+                            from_date=timestamp,
+                            to_date=timestamp,
+                            size=len(data),
+                            events=[dict(event_id=event_id, timestamp=timestamp, offset=0, size=len(data))])
+            self._upload_journal(_Journal(journal_id, manifest, io.BytesIO(data)), is_new=True)
+            return Event(event_id, date, data)
+
+    def delete(self, event_id):
```
Comment (Bento007) [inline flashflood/__init__.py]: missing typing

Comment (xbrianh) [reply flashflood/__init__.py]: Type are needed throughout this repo. That will be addressed in a future PR.

File: flashflood/util.py [hunk 3]
```diff
@@ -66,3 +70,33 @@ def overlaps(self, *args, **kwargs):
 
     def contains(self, *args, **kwargs):
         return False
+
+def delete_keys(bucket, keys):
+    def _delete(keys_to_delete):
+        assert 1000 >= len(keys_to_delete)
```
Comment (Bento007) [inline flashflood/util.py]: is there a case where this triggered? Otherwise it seems excessive.

Comment (Bento007) [reply flashflood/util.py]: Pull out the 1000 into a variable which describes it.

Comment (xbrianh) [reply flashflood/util.py]: That's the maximum number of keys allowed in an S3 batch delete request.

Will fix.

**Pull request merged**